plugins/thumbnails/server/__init__.py

```diff
@@ -53,13 +54,46 @@ def removeThumbnailLink(event):
             model.save(resource, validate=False)
 
 
+def _onUpload(event):
+    """
+    Thumbnail creation can be requested on file upload by passing a reference field
+    that is a JSON object of the following form:
+
+        {
+          "thumbnail": {
+            "width": 123,
+            "height": 123,
+            "crop": True
+          }
+        }
+
+    At least one of ``width`` or ``height`` must be passed. The ``crop`` parameter is optional.
+    """
+    try:
+        ref = json.loads(event.info.get('reference'))
+    except (ValueError, TypeError):
+        return
+
+    if isinstance(ref, dict) and isinstance(ref.get('thumbnail'), dict):
+        width = max(0, ref['thumbnail'].get('width', 0))
+        height = max(0, ref['thumbnail'].get('height', 0))
+
+        if not width and not height:
+            return
+
+        file = event.info['file']
+        item = ModelImporter.model('item').load(file['itemId'], force=True)
+        utils.scheduleThumbnailJob(
+            file=file, attachToType='item', attachToId=item['_id'], user=event.info['currentUser'],
+            width=width, height=height, crop=ref['thumbnail'].get('crop', True))
```

Comment:
I chose to extract the pythonic truthiness value of the "crop" field, which is not the same as validation but is probably good enough.